File: .claude/hooks/explore.py
#!/usr/bin/env python3
"""PostToolUse hook: records exploration events (Read, Glob, Grep, LS, WebSearch, WebFetch)."""

import json
import os
import sys
import urllib.request

API_BASE    = f"http://127.0.0.1:{os.environ.get('MONITOR_PORT', '3847')}"
PROJECT_DIR = os.environ.get("CLAUDE_PROJECT_DIR", os.getcwd())
MAX_PATH_LEN = 300
OPENCODE_RUNTIME = bool(os.environ.get("OPENCODE") or os.environ.get("OPENCODE_CLIENT"))
CLAUDE_RUNTIME = bool(os.environ.get("CLAUDE_PROJECT_DIR")) and not OPENCODE_RUNTIME


def _rel(path: str) -> str:
    if path.startswith(PROJECT_DIR):
        return path[len(PROJECT_DIR):].lstrip("/")
    return path


def _post(path: str, body: dict) -> None:
    req = urllib.request.Request(
        f"{API_BASE}{path}",
        data=json.dumps(body).encode(),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    urllib.request.urlopen(req, timeout=2)


def _get_ids(cc_session_id: str) -> tuple[str, str] | None:
    req = urllib.request.Request(
        f"{API_BASE}/api/cc-session-ensure",
        data=json.dumps({
            "ccSessionId":   cc_session_id,
            "title":         f"Claude Code — {os.path.basename(PROJECT_DIR)}",
            "workspacePath": PROJECT_DIR,
        }).encode(),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=2) as resp:
        data = json.loads(resp.read())
    return data["taskId"], data["sessionId"]


def main() -> None:
    if not CLAUDE_RUNTIME:
        return

    try:
        event         = json.load(sys.stdin)
        tool_name     = event.get("tool_name", "")
        tool_input    = event.get("tool_input", {})
        cc_session_id = (event.get("session_id") or "").strip()
    except Exception:
        return

    if not cc_session_id:
        return

    try:
        task_id, session_id = _get_ids(cc_session_id)
    except Exception:
        return

    if tool_name == "Read":
        fp    = tool_input.get("file_path", "")
        rel   = _rel(fp)
        title = f"Read: {os.path.basename(rel)}"
        body  = f"Reading {rel}"
        paths = [fp] if fp else []
    elif tool_name == "Glob":
        pattern = tool_input.get("pattern", "")
        title   = f"Glob: {pattern}"
        body    = f"Searching for files matching: {pattern}"
        paths   = []
    elif tool_name == "Grep":
        pattern = tool_input.get("pattern", "")
        path    = tool_input.get("path", "")
        rel     = _rel(path) if path else ""
        title   = f"Grep: {pattern[:60]}"
        body    = f"Searching for '{pattern}'" + (f" in {rel}" if rel else "")
        paths   = [path] if path else []
    elif tool_name == "LS":
        path  = tool_input.get("path", "")
        rel   = _rel(path) if path else "."
        title = f"LS: {rel}"
        body  = f"Listing directory: {rel}"
        paths = [path] if path else []
    elif tool_name in ("WebSearch", "WebFetch"):
        query = tool_input.get("query") or tool_input.get("url") or ""
        title = f"{tool_name}: {str(query)[:60]}"
        body  = f"Web lookup: {str(query)[:200]}"
        paths = []
    else:
        title = f"Explore: {tool_name}"
        body  = f"Used {tool_name} to explore"
        paths = []

    try:
        _post("/api/explore", {
            "taskId":    task_id,
            "sessionId": session_id,
            "toolName":  tool_name,
            "title":     title,
            "body":      body,
            "filePaths": [p[:MAX_PATH_LEN] for p in paths if p],
            "metadata":  {"toolInput": {k: str(v)[:200] for k, v in tool_input.items()}},
        })
    except Exception:
        pass


main()

File: packages/hook-plugin/hooks/PostToolUse/Explore.ts
/**
 * Claude Code Hook: PostToolUse — matcher: "Read|Glob|Grep|WebSearch|WebFetch"
 *
 * Fires after any exploration tool succeeds. Does not fire on failures.
 *
 * Stdin payload fields (ref: https://code.claude.com/docs/en/hooks#posttooluse):
 *   session_id       string  — unique session identifier
 *   hook_event_name  string  — "PostToolUse"
 *   tool_name        string  — "Read" | "Glob" | "Grep" | "WebSearch" | "WebFetch"
 *   tool_input       object  — tool-specific input (see below)
 *   tool_response    any     — tool output (potentially large; not used here)
 *   tool_use_id      string  — unique ID for this tool invocation
 *   cwd              string  — current working directory
 *   transcript_path  string  — path to the session transcript JSONL
 *   permission_mode  string  — current permission mode
 *   agent_id         string? — set when inside a subagent
 *
 * Tool-specific tool_input fields:
 *   Read:      { file_path: string }
 *   Glob:      { pattern: string, path?: string }
 *   Grep:      { pattern: string, path?: string, glob?: string }
 *   WebSearch: { query: string }
 *   WebFetch:  { url: string, prompt?: string }
 *
 * Blocking: PostToolUse cannot block (exit 2 shows stderr but execution continues).
 *
 * This handler classifies the explore operation semantically and posts
 * an /api/explore event to the Agent Tracer monitor.
 */
import * as path from "node:path";
import { relativeProjectPath } from "../util/paths.js";
import { getAgentContext, getSessionId, getToolInput, getToolName, getToolUseId, stringifyToolInput, toTrimmedString } from "../util/utils.js";
import { postJson, readStdinJson } from "../lib/transport.js";
import { resolveEventSessionIds } from "../lib/subagent-session.js";
import { hookLog, hookLogPayload } from "../lib/hook-log.js";
import { inferExploreSemantic } from "../classification/explore-semantic.js";
import { buildSemanticMetadata } from "../classification/command-semantic.js";

const MAX_PATH_LENGTH = 300;

async function main(): Promise<void> {
    const payload = await readStdinJson();
    hookLogPayload("PostToolUse/Explore", payload);
    const sessionId = getSessionId(payload);
    const { agentId, agentType } = getAgentContext(payload);
    const toolName = getToolName(payload);
    const toolInput = getToolInput(payload);
    hookLog("PostToolUse/Explore", "fired", { toolName, sessionId: sessionId || "(none)" });

    if (!sessionId) {
        hookLog("PostToolUse/Explore", "skipped — no sessionId");
        return;
    }

    const ids = await resolveEventSessionIds(sessionId, agentId, agentType);
    const toolUseId = getToolUseId(payload);
    let title = `Explore: ${toolName}`;
    let body = `Used ${toolName} to explore`;
    let filePaths: string[] = [];

    if (toolName === "Read") {
        const filePath = toTrimmedString(toolInput.file_path);
        const relPath = relativeProjectPath(filePath);
        title = `Read: ${path.basename(relPath)}`;
        body = `Reading ${relPath}`;
        filePaths = filePath ? [filePath] : [];
    } else if (toolName === "Glob") {
        const pattern = toTrimmedString(toolInput.pattern);
        title = `Glob: ${pattern}`;
        body = `Searching for files matching: ${pattern}`;
    } else if (toolName === "Grep") {
        const pattern = toTrimmedString(toolInput.pattern);
        const searchPath = toTrimmedString(toolInput.path);
        const relPath = searchPath ? relativeProjectPath(searchPath) : "";
        title = `Grep: ${pattern.slice(0, 60)}`;
        body = `Searching for '${pattern}'${relPath ? ` in ${relPath}` : ""}`;
        filePaths = searchPath ? [searchPath] : [];
    } else if (toolName === "WebSearch" || toolName === "WebFetch") {
        const query = toTrimmedString(toolInput.query) || toTrimmedString(toolInput.url);
        title = `${toolName}: ${query.slice(0, 60)}`;
        body = `Web lookup: ${query}`;
    }

    const semantic = inferExploreSemantic(toolName, toolInput);
    const isWebTool = toolName === "WebSearch" || toolName === "WebFetch";
    const webQuery = isWebTool
        ? (toTrimmedString(toolInput.query) || toTrimmedString(toolInput.url)).slice(0, MAX_PATH_LENGTH)
        : "";

    await postJson("/ingest/v1/events", {
        events: [{
            kind: "tool.used",
            lane: "exploration",
            taskId: ids.taskId,
            sessionId: ids.sessionId,
            toolName,
            title,
            body,
            filePaths: filePaths.map((fp) => fp.slice(0, MAX_PATH_LENGTH)),
            metadata: {
                ...buildSemanticMetadata(semantic),
                toolInput: stringifyToolInput(toolInput),
                ...(isWebTool && webQuery ? { webUrls: [webQuery] } : {}),
                ...(toolUseId ? { toolUseId } : {})
            }
        }]
    });
    hookLog("PostToolUse/Explore", "explore posted", { toolName, title });
}

void main().catch((err: unknown) => {
    hookLog("PostToolUse/Explore", "ERROR", { error: String(err) });
});
